services/mpa/server/server.go

```diff
@@ -0,0 +1,291 @@
+		if act.lastModified.Before(staleTime) {
+			delete(s.actions, id)
+		}
+	}
+
+}
+
+func (s *server) Store(ctx context.Context, in *mpa.StoreRequest) (*mpa.StoreResponse, error) {
+	var justification string
+	if md, found := metadata.FromIncomingContext(ctx); found && len(md[rpcauth.ReqJustKey]) > 0 {
+		justification = md[rpcauth.ReqJustKey][0]
+	}
+
+	p, ok := callerIdentity(ctx)
+	if !ok || p == nil {
+		return nil, status.Error(codes.FailedPrecondition, "unable to determine caller's identity")
+	}
+
+	action := &mpa.Action{
+		User:          p.ID,
+		Justification: justification,
+		Method:        in.Method,
+		Message:       in.Message,
+	}
+	id, err := actionId(action)
+	if err != nil {
+		return nil, err
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Time to clear out excessive approvals!
+	for len(s.actions) >= maxMPAApprovals {
+		var oldestID string
+		oldestTime := time.Now()
+		for id, act := range s.actions {
+			if act.lastModified.Before((oldestTime)) {
+				oldestID = id
+				oldestTime = act.lastModified
+			}
+		}
+		delete(s.actions, oldestID)
+	}
+
+	act, ok := s.actions[id]
+	if !ok {
+		act = &storedAction{
+			action:       action,
+			approved:     make(chan struct{}),
+			lastModified: time.Now(),
+		}
+		s.actions[id] = act
+	}
+	return &mpa.StoreResponse{
+		Id:       id,
+		Action:   action,
+		Approver: act.approvers,
+	}, nil
+}
+
+func containsPrincipal(principals []*mpa.Principal, p *rpcauth.PrincipalAuthInput) bool {
+	for _, s := range principals {
+		if s.Id == p.ID && reflect.DeepEqual(s.Groups, p.Groups) {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *server) Approve(ctx context.Context, in *mpa.ApproveRequest) (*mpa.ApproveResponse, error) {
+	p, ok := callerIdentity(ctx)
+	if !ok {
+		return nil, status.Error(codes.FailedPrecondition, "unable to determine caller's identity")
+	}
+	id, err := actionId(in.Action)
+	if err != nil {
+		return nil, err
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	act, ok := s.actions[id]
+	if !ok {
+		return nil, status.Error(codes.NotFound, "MPA request with provided input not found")
+	}
+	if act.action.User == p.ID {
+		return nil, status.Error(codes.InvalidArgument, "MPA requests cannot be approved by their requestor")
+	}
+	act.lastModified = time.Now()
+	// Only add the approver if it's new compared to existing approvals
+	if !containsPrincipal(act.approvers, p) {
+		act.approvers = append(act.approvers, &mpa.Principal{
+			Id:     p.ID,
+			Groups: p.Groups,
+		})
+		// The first approval lets any WaitForApproval calls finish immediately.
+		if len(act.approvers) == 1 {
+			close(act.approved)
+		}
+	}
+	return &mpa.ApproveResponse{}, nil
+}
+func (s *server) WaitForApproval(ctx context.Context, in *mpa.WaitForApprovalRequest) (*mpa.WaitForApprovalResponse, error) {
+	s.mu.Lock()
+	act, ok := s.actions[in.Id]
+	if !ok {
+		return nil, status.Error(codes.NotFound, "MPA request not found")
+	}
+	s.mu.Unlock()
+	select {
+	case <-act.approved:
```

Comment:
one minor enhancement you might make here is to periodically recheck that the action is still valid, in case the client has set a very long deadline

e.g
```go
for {
        s.mu.Lock()
	act, ok := s.actions[in.Id]
	if !ok {
		return nil, status.Error(codes.NotFound, "MPA request not found")
	}
	s.mu.Unlock()
        timer := time.NewTimer(5 * time.Minute)
        select {
          case <-act.approved:
            return &mpa.WaitForApprovalResponse{}, nil
          case <-ctx.Done()
             return nil, ctx.Err()
          case <-timer.C:
             // go around the loop again
        }
}
```

Reply:
Good idea, done. I'm doing once-a-minute checks because checking should be cheap (single lock/unlock)